README.md
Recommendation
==============

**Collaborative filtering for recommender system**

The gem is intended to implement collaborative filtering alone. (not use KVS.)


What is Recommendation
----------------------

See also.

+ http://en.wikipedia.org/wiki/Recommender_system


What is Collaborative filtering
-------------------------------

See also.

+ http://en.wikipedia.org/wiki/Collaborative_filtering


In reference
------------

+ http://www.amazon.co.jp/Programming-Collective-Intelligence-Building-Applications/dp/0596529325


Tutorial
--------

See demo.rb and spec files.


ChangeLog
---------

See doc/ChangeLog.


Developers
----------

See doc/AUTHORS.


Author
------

**774**

+ http://id774.net
+ http://github.com/id774


Copyright and license
---------------------

See the file doc/LICENSE.




doc/ChangeLog
=== 0.3.0 / 2013-12-13

* Improvement spec (for RSpec 2.14.x).

* Re-construction filenames.


=== 0.2.0 / 2013-10-19

* Change output format to Array.


=== 0.1.3 / 2013-10-18

* First release.




spec/spec_helper.rb
# -*- coding: utf-8 -*-

APP_ROOT = File.expand_path(File.join(File.dirname(__FILE__), ".."))
$LOAD_PATH.unshift APP_ROOT
$LOAD_PATH.unshift File.join(APP_ROOT)
$LOAD_PATH.unshift File.join(APP_ROOT, 'lib')

require 'recommendation'

if ENV['COVERAGE'] == 'on'
  require 'simplecov'
  require 'simplecov-rcov'
  SimpleCov.formatter = SimpleCov::Formatter::RcovFormatter

  SimpleCov.start do
    add_filter "spec"
    add_filter "vendor"
  end
end

unless /^1\.9\./ =~ RUBY_VERSION
  require 'rspec'
end

RSpec.configure do |config|
end


spec/lib/recommendation_spec.rb
# -*- coding: utf-8 -*-

require File.dirname(__FILE__) + '/../spec_helper'

describe Recommendation do
  context "VERSION" do
    subject { Recommendation::VERSION }

    it { expect(subject).to eq "0.3.0" }
  end
end


spec/lib/recommendation/engine_spec.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require File.dirname(__FILE__) + '/../../spec_helper'

describe 'Recommendation::Engine' do
  context '#recommendation' do

    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      engine.recommendation(supervisor.table, new_comer.keys[0])
    }

    let(:expected) {
      [
        ["The Night Listener", 3.3477895267131017],
        ["Lady in the Water", 2.8325499182641614],
        ["Just My Luck", 2.530980703765565]
      ]
    }

    it 'should be suggesting interesting products' do
      expect(subject).to eq expected
    end
  end

  context '#top_matches' do

    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      engine.top_matches(supervisor.table, new_comer.keys[0])
    }

    let(:expected) {
      [
        ["Lisa Rose", 0.9912407071619299],
        ["Mick LaSalle", 0.9244734516419049],
        ["Claudia Puig", 0.8934051474415647],
        ["Jack Matthews", 0.66284898035987],
        ["Gene Seymour", 0.38124642583151164]
      ]
    }

    it 'should be finding similar users' do
      expect(subject).to eq expected
    end
  end

  context 'reversed critics' do

    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      movies = supervisor.transform_table
      engine.top_matches(movies, new_comer.values[0].keys[2])
    }

    let(:expected) {
      [
        ["You, Me and Dupree", 0.6579516949597695],
        ["Lady in the Water", 0.4879500364742689],
        ["Snake on the Plane", 0.11180339887498941],
        ["The Night Listener", -0.1798471947990544],
        ["Just My Luck", -0.42289003161103106]
      ]
    }

    it 'should be found similar items' do
      expect(subject).to eq expected
    end
  end

  context 'recommendation for the unexisting user' do

    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      engine.recommendation(supervisor.table, 'hoge')
    }

    let(:expected) { [] }

    it 'should return empty array' do
      expect(subject).to eq expected
    end
  end

  context 'top_matches for the unexisting item' do
    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      engine.top_matches(supervisor.table, 'fuga')
    }

    let(:expected) {
      [
        ["Toby", 0],
        ["Mick LaSalle", 0],
        ["Michael Phillips", 0],
        ["Lisa Rose", 0],
        ["Jack Matthews", 0]
      ]
    }

    it 'should return all zero' do
      expect(subject).to eq expected
    end
  end
end

describe 'Recommendation::Supervisor' do
  context '#transform_table ' do
    subject {
      supervisor = Recommendation::Supervisor.new(visitors)
      supervisor.train(new_comer)
      engine = Recommendation::Engine.new
      supervisor.transform_table
    }

    it 'should return reversed critics' do
      expect(subject).to eq reversed_critics
    end
  end
end

def new_comer
  {
    'Toby' => {
      'Snake on the Plane' => 4.5,
      'You, Me and Dupree' => 1.0,
      'Superman Returns'   => 4.0
    }
  }
end

def visitors
  {
    'Lisa Rose' => {
      'Lady in the Water'  => 2.5,
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 3.0,
      'Superman Returns'   => 3.5,
      'You, Me and Dupree' => 2.5,
      'The Night Listener' => 3.0
    },

    'Gene Seymour' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 1.5,
      'Superman Returns'   => 5.0,
      'The Night Listener' => 3.0,
      'You, Me and Dupree' => 3.5
    },

    'Michael Phillips' => {
      'Lady in the Water'  => 2.5,
      'Snake on the Plane' => 3.0,
      'Superman Returns'   => 3.5,
      'The Night Listener' => 4.0
    },

    'Claudia Puig' => {
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 3.0,
      'The Night Listener' => 4.5,
      'Superman Returns'   => 4.0,
      'You, Me and Dupree' => 2.5
    },

    'Mick LaSalle' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 4.0,
      'Just My Luck'       => 2.0,
      'Superman Returns'   => 3.0,
      'The Night Listener' => 3.0,
      'You, Me and Dupree' => 2.0
    },

    'Jack Matthews' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 4.0,
      'The Night Listener' => 3.0,
      'Superman Returns'   => 5.0,
      'You, Me and Dupree' => 3.5
    }
  }
end

def reversed_critics
  {
    "Lady in the Water" => {
      "Lisa Rose"        => 2.5,
      "Gene Seymour"     => 3.0,
      "Michael Phillips" => 2.5,
      "Mick LaSalle"     => 3.0,
      "Jack Matthews"    => 3.0
    },

    "Snake on the Plane" => {
      "Lisa Rose"        => 3.5,
      "Gene Seymour"     => 3.5,
      "Michael Phillips" => 3.0,
      "Claudia Puig"     => 3.5,
      "Mick LaSalle"     => 4.0,
      "Jack Matthews"    => 4.0,
      "Toby"             => 4.5
    },

    "Just My Luck" => {
      "Lisa Rose"        => 3.0,
      "Gene Seymour"     => 1.5,
      "Claudia Puig"     => 3.0,
      "Mick LaSalle"     => 2.0
    },

    "Superman Returns" => {
      "Lisa Rose"        => 3.5,
      "Gene Seymour"     => 5.0,
      "Michael Phillips" => 3.5,
      "Claudia Puig"     => 4.0,
      "Mick LaSalle"     => 3.0,
      "Jack Matthews"    => 5.0,
      "Toby"             => 4.0
    },

    "You, Me and Dupree" => {
      "Lisa Rose"        => 2.5,
      "Gene Seymour"     => 3.5,
      "Claudia Puig"     => 2.5,
      "Mick LaSalle"     => 2.0,
      "Jack Matthews"    => 3.5,
      "Toby"             => 1.0
    },

    "The Night Listener" => {
      "Lisa Rose"        => 3.0,
      "Gene Seymour"     => 3.0,
      "Michael Phillips" => 4.0,
      "Claudia Puig"     => 4.5,
      "Mick LaSalle"     => 3.0,
      "Jack Matthews"    => 3.0
    }
  }
end


lib/recommendation/engine.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

module Recommendation
  class Engine
    def recommendation(table, user, similarity=:sim_pearson)
      totals_h = Hash.new(0)
      sim_sums_h = Hash.new(0)
      table.each do |other, val|
        next if other == user
        sim = __send__(similarity, table, user, other)
        next if sim <= 0
        table[other].each do |item, val|
          if !table[user].keys.include?(item) || table[user][item] == 0
            totals_h[item] += table[other][item]*sim
            sim_sums_h[item] += sim
          end
        end
      end

      rankings = Hash.new
      totals_h.each do |item, total|
        rankings[item] = total/sim_sums_h[item]
      end

      rankings.sort_by{|k, v| -v}
    end

    def top_matches(table, user, n=5, similarity=:sim_pearson)
      scores = Array.new
      table.each do |key, value|
        if key != user
          scores << [__send__(similarity, table, user,key), key]
        end
      end

      result = Array.new
      scores.sort.reverse[0,n].each do |k, v|
        result << [v, k]
      end
      result
    end

    private

    def sim_pearson(table, user1, user2)
      shared_items_a = shared_items_a(table, user1, user2)

      n = shared_items_a.size
      return 0 if n == 0

      sum1 = shared_items_a.inject(0) {|result, si|
        result + table[user1][si]
      }
      sum2 = shared_items_a.inject(0) {|result, si|
        result + table[user2][si]
      }

      sum1_sq = shared_items_a.inject(0) {|result, si|
        result + table[user1][si]**2
      }
      sum2_sq = shared_items_a.inject(0) {|result, si|
        result + table[user2][si]**2
      }

      sum_products = shared_items_a.inject(0) {|result, si|
        result + table[user1][si]*table[user2][si]
      }

      num = sum_products - (sum1*sum2/n)
      den = Math.sqrt((sum1_sq - sum1**2/n)*(sum2_sq - sum2**2/n))
      return 0 if den == 0
      return num/den
    end

    def shared_items(table, user1, user2)
      shared_items_h = Hash.new
      table[user1].each do |k, v|
        shared_items_h[k] = 1 if table[user2].include?(k)
      end
      shared_items_h
    end

    def shared_items_a(table, user1, user2)
      table[user1].nil? ? [] : table[user1].keys & table[user2].keys
    end
  end
end


lib/recommendation/recommender.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

module Recommendation
  class Recommender
    attr_accessor :table

    def initialize(params = {})
      @table = params
    end

    def train(params = {})
      @table.merge!(params)
    end

    def transform_table
      new_table = {}
      @table.each do |key, value|
        value.each do |new_key, new_value|
          new_table[new_key] ||= Hash.new
          new_table[new_key][key] = new_value
        end
      end
      new_table
    end

    class << self
      def recommendation(table, user, similarity=:sim_pearson)
        totals_h = Hash.new(0)
        sim_sums_h = Hash.new(0)
        table.each do |other, val|
          next if other == user
          sim = __send__(similarity, table, user, other)
          next if sim <= 0
          table[other].each do |item, val|
            if !table[user].keys.include?(item) || table[user][item] == 0
              totals_h[item] += table[other][item]*sim
              sim_sums_h[item] += sim
            end
          end
        end

        rankings = Hash.new
        totals_h.each do |item, total|
          rankings[item] = total/sim_sums_h[item]
        end

        rankings.sort_by{|k, v| -v}
      end

      def top_matches(table, user, n=5, similarity=:sim_pearson)
        scores = Array.new
        table.each do |key, value|
          if key != user
            scores << [__send__(similarity, table, user,key), key]
          end
        end

        result = Array.new
        scores.sort.reverse[0,n].each do |k, v|
          result << [v, k]
        end
        result
      end

      private

      def sim_pearson(table, user1, user2)
        shared_items_a = shared_items_a(table, user1, user2)

        n = shared_items_a.size
        return 0 if n == 0

        sum1 = shared_items_a.inject(0) {|result, si|
          result + table[user1][si]
        }
        sum2 = shared_items_a.inject(0) {|result, si|
          result + table[user2][si]
        }

        sum1_sq = shared_items_a.inject(0) {|result, si|
          result + table[user1][si]**2
        }
        sum2_sq = shared_items_a.inject(0) {|result, si|
          result + table[user2][si]**2
        }

        sum_products = shared_items_a.inject(0) {|result, si|
          result + table[user1][si]*table[user2][si]
        }

        num = sum_products - (sum1*sum2/n)
        den = Math.sqrt((sum1_sq - sum1**2/n)*(sum2_sq - sum2**2/n))
        return 0 if den == 0
        return num/den
      end

      def shared_items(table, user1, user2)
        shared_items_h = Hash.new
        table[user1].each do |k, v|
          shared_items_h[k] = 1 if table[user2].include?(k)
        end
        shared_items_h
      end

      def shared_items_a(table, user1, user2)
        table[user1].nil? ? [] : table[user1].keys & table[user2].keys
      end
    end
  end
end


demo.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

$:.unshift File.join(File.dirname(__FILE__))

require 'lib/recommendation'

def demo
  recommender = Recommendation::Recommender.new(visitors)
  recommender.train(new_comer)

  p Recommendation::Recommender.recommendation(recommender.table, 'Toby') # => [["The Night Listener", 3.3477895267131017], ["Lady in the Water", 2.8325499182641614], ["Just My Luck", 2.530980703765565]]

  p Recommendation::Recommender.top_matches(recommender.table, 'Toby') # => [["Lisa Rose", 0.9912407071619299], ["Mick LaSalle", 0.9244734516419049], ["Claudia Puig", 0.8934051474415647], ["Jack Matthews", 0.66284898035987], ["Gene Seymour", 0.38124642583151164]]

  movies = recommender.transform_table
  p Recommendation::Recommender.top_matches(movies, 'Superman Returns') # => [["You, Me and Dupree", 0.6579516949597695], ["Lady in the Water", 0.4879500364742689], ["Snake on the Plane", 0.11180339887498941], ["The Night Listener", -0.1798471947990544], ["Just My Luck", -0.42289003161103106]]

end

def new_comer
  {
    'Toby' => {
      'Snake on the Plane' => 4.5,
      'You, Me and Dupree' => 1.0,
      'Superman Returns'   => 4.0
    }
  }
end

def visitors
  {
    'Lisa Rose' => {
      'Lady in the Water'  => 2.5,
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 3.0,
      'Superman Returns'   => 3.5,
      'You, Me and Dupree' => 2.5,
      'The Night Listener' => 3.0
    },

    'Gene Seymour' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 1.5,
      'Superman Returns'   => 5.0,
      'The Night Listener' => 3.0,
      'You, Me and Dupree' => 3.5
    },

    'Michael Phillips' => {
      'Lady in the Water'  => 2.5,
      'Snake on the Plane' => 3.0,
      'Superman Returns'   => 3.5,
      'The Night Listener' => 4.0
    },

    'Claudia Puig' => {
      'Snake on the Plane' => 3.5,
      'Just My Luck'       => 3.0,
      'The Night Listener' => 4.5,
      'Superman Returns'   => 4.0,
      'You, Me and Dupree' => 2.5
    },

    'Mick LaSalle' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 4.0,
      'Just My Luck'       => 2.0,
      'Superman Returns'   => 3.0,
      'The Night Listener' => 3.0,
      'You, Me and Dupree' => 2.0
    },

    'Jack Matthews' => {
      'Lady in the Water'  => 3.0,
      'Snake on the Plane' => 4.0,
      'The Night Listener' => 3.0,
      'Superman Returns'   => 5.0,
      'You, Me and Dupree' => 3.5
    }
  }
end

demo
